core/utils/checksum.h

```diff
@@ -109,9 +110,11 @@ static inline uint32_t CalculateSum(const void *buf, size_t len) {
   // Reduce 64-bit unsigned integer to 32-bit unsigned integer
   // Carry may happens, but no need to complete reduce here
   sum64 = (sum64 >> 32) + (sum64 & 0xFFFFFFFF);
+#endif
 
   // Repeat 16-bit one's complement sum (at sum64)
-  const uint16_t *buf16 = reinterpret_cast<const uint16_t *>(buf64);
+  typedef uint16_t __attribute__((__may_alias__)) u16_p;
```

Comment:
Have you run the unit tests? It was not a compile time issue but a run time issue. It seems when there is optimization -O2 or greater the optimizer presumes strict aliasing. Note that this is the same technique used rte_raw_cksum(). I suspect when the remainder of the function is ifdef'd out (as is the case in 32 bit mode) the optimizer kicks in and the strict aliasing issue shows up (I don't think you see the optimizer effects on this section of the code in 64 bit mode).